sw/device/tests/dif/dif_clkmgr_smoketest.c

```diff
@@ -0,0 +1,95 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+#include "sw/device/lib/base/memory.h"
+#include "sw/device/lib/dif/dif_clkmgr.h"
+#include "sw/device/lib/testing/check.h"
+#include "sw/device/lib/testing/test_main.h"
+
+#include "hw/top_earlgrey/sw/autogen/top_earlgrey.h"  // Generated.
+
+const test_config_t kTestConfig = {0};
+
+/**
+ * Test that all 'gateable' clocks, directly controlled by software,
+ * can be enabled and disabled.
+ */
+static void test_gateable_clocks(const dif_clkmgr_t *clkmgr) {
+  const dif_clkmgr_gateable_clock_t clocks[] = {
+      kTopEarlgreyGateableClocksIoDiv4Peri,
+      kTopEarlgreyGateableClocksUsbPeri,
+  };
+
+  for (int i = 0; i < ARRAYSIZE(clocks); ++i) {
+    dif_clkmgr_gateable_clock_t clock = clocks[i];
+    bool enabled;
+    CHECK(dif_clkmgr_gateable_clock_get_enabled(clkmgr, clock, &enabled) ==
+          kDifClkmgrOk);
+
+    // Toggle the enable twice so that it ends up in its original state.
+    for (int j = 0; j < 2; ++j) {
+      bool expected = !enabled;
+      CHECK(dif_clkmgr_gateable_clock_set_enabled(
+                clkmgr, clock,
+                expected ? kDifClkmgrToggleEnabled
+                         : kDifClkmgrToggleDisabled) == kDifClkmgrOk);
+      CHECK(dif_clkmgr_gateable_clock_get_enabled(clkmgr, clock, &enabled) ==
+            kDifClkmgrOk);
+      CHECK(enabled == expected);
+    }
+  }
+}
+
+/**
+ * Test that all 'hintable' clocks, indirectly controlled by software,
+ * can have their hint toggled and status checked.
+ */
+void test_hintable_clocks(const dif_clkmgr_t *clkmgr) {
+  const dif_clkmgr_hintable_clock_t clocks[] = {
+      kTopEarlgreyHintableClocksMainAes,
+      kTopEarlgreyHintableClocksMainHmac,
+      kTopEarlgreyHintableClocksMainKmac,
+      kTopEarlgreyHintableClocksMainOtbn,
+  };
+
+  for (int i = 0; i < ARRAYSIZE(clocks); ++i) {
+    dif_clkmgr_hintable_clock_t clock = clocks[i];
+    bool enabled;
+    CHECK(dif_clkmgr_hintable_clock_get_hint(clkmgr, clock, &enabled) ==
+          kDifClkmgrOk);
+
+    // Toggle the hint twice so that it ends up in its original state.
+    for (int j = 0; j < 2; ++j) {
+      bool expected = !enabled;
+      CHECK(dif_clkmgr_hintable_clock_set_hint(
+                clkmgr, clock,
+                expected ? kDifClkmgrToggleEnabled
+                         : kDifClkmgrToggleDisabled) == kDifClkmgrOk);
+      CHECK(dif_clkmgr_hintable_clock_get_hint(clkmgr, clock, &enabled) ==
+            kDifClkmgrOk);
+      CHECK(enabled == expected);
+    }
+
+    // Just check that a call to get enabled succeeds.
+    // Software has no direct control over this value.
+    // TODO: or does it? If hint is enabled then this should be enabled too...
```

Comment:
nit: It would be nice to resolve this if possible before this change lands.

Reply:
My understanding is that setting the hint to 'enabled' should always enable the clock. I think that means the status will always be enabled if the hint is enabled. I've changed the test to reflect that.